packages/core/data-transfer/src/strapi/providers/local-source/assets.ts

```diff
@@ -1,32 +1,100 @@
 import { join } from 'path';
-import { readdir, stat, createReadStream } from 'fs-extra';
-import { Duplex } from 'stream';
+import https from 'https';
+import { stat, createReadStream, ReadStream } from 'fs-extra';
+import { Duplex, PassThrough } from 'stream';
 
-import type { IAsset } from '../../../../types';
+import type { IAsset, IFile } from '../../../../types';
 
-const IGNORED_FILES = ['.gitkeep'];
+function getFileStream(filepath: string, isLocal = false): PassThrough | ReadStream {
+  if (isLocal) {
+    return createReadStream(filepath);
+  }
 
+  const readableStream = new PassThrough();
+
+  https
+    .get(filepath, (res) => {
+      if (res.statusCode !== 200) {
+        readableStream.emit(
+          'error',
+          new Error(`Request failed with status code ${res.statusCode}`)
+        );
+        return;
+      }
+
+      res.pipe(readableStream);
+    })
+    .on('error', (error) => {
+      readableStream.emit('error', error);
+    });
+
+  return readableStream;
+}
+
+function getFileStats(filepath: string, isLocal = false): Promise<{ size: number }> {
+  if (isLocal) {
+    return stat(filepath);
+  }
+  return new Promise((resolve, reject) => {
+    https
+      .get(filepath, (res) => {
+        if (res.statusCode !== 200) {
+          reject(new Error(`Request failed with status code ${res.statusCode}`));
+          return;
+        }
+
+        const contentLength = res.headers['content-length'];
+        const stats = {
+          size: contentLength ? parseInt(contentLength, 10) : 0,
+        };
+
+        resolve(stats);
+      })
+      .on('error', (error) => {
+        reject(error);
+      });
+  });
+}
 /**
  * Generate and consume assets streams in order to stream each file individually
  */
 export const createAssetsStream = (strapi: Strapi.Strapi): Duplex => {
-  const assetsDirectory = join(strapi.dirs.static.public, 'uploads');
-
   const generator: () => AsyncGenerator<IAsset, void> = async function* () {
-    const files = await readdir(assetsDirectory);
-    const validFiles = files.filter((file) => !IGNORED_FILES.includes(file));
+    const files: IFile[] = await strapi.query('plugin::upload.file').findMany();
```

Comment:
Instead of querying all the files at once, why don't we stream them?

Reply:
Here's an old example of what I've done to make a POC for the cloud. It's not using the metadata, but maybe you can reuse some stuff about the streaming
```ts
import { Duplex, Readable } from 'stream';
import * as axios from 'axios';
​
import type { IAsset } from '../../../../types';
​
/**
 * Generate and consume assets streams in order to stream each file individually
 */
export const createAssetsStream = (strapi: Strapi.Strapi): Duplex => {
  const generator: () => AsyncGenerator<IAsset, void> = async function* () {
    const stream: Readable = strapi.db
      // Create a query builder instance (default type is 'select')
      .queryBuilder('plugin::upload.file')
      // Fetch all columns
      .select('*')
      // Get a readable stream
      .stream();
​
    for await (const file of stream) {
      const head = await (axios as any).head(file.url, { responseType: 'blob' });
      const res = await (axios as any).get(file.url, { responseType: 'stream' });
​
      yield {
        filename: file.name,
        filepath: `${file.name}.${file.ext}`,
        stream: res.data,
        stats: { size: parseInt(head.headers['content-length'], 10) },
      };
    }
  };
​
  return Duplex.from(generator());
};
```